<a href="https://colab.research.google.com/github/Sharma-Shivam2021/credit-default-prediction/blob/main/notebooks/02_preprocessing_and_baseline.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!git config user.email "2571sharmashivam@gmail.com"
!git config user.name "Shivam Sharma"

fatal: not in a git directory
fatal: not in a git directory


In [61]:
!git status

On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean


In [62]:
!git add .

In [63]:
!git commit -m "Updated notebook 2"

On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean


In [64]:
import getpass

token = getpass.getpass("Give your token here: ")

username = "Sharma-Shivam2021"
!git remote set-url origin https://{username}:{token}@github.com/{username}/credit-default-prediction.git
!git push origin main

Give your token here: ··········
Everything up-to-date


In [6]:
!git clone https://github.com/Sharma-Shivam2021/credit-default-prediction.git
%cd credit-default-prediction

fatal: destination path 'credit-default-prediction' already exists and is not an empty directory.
/content/credit-default-prediction


In [7]:
%cd /content/credit-default-prediction

/content/credit-default-prediction


In [8]:
%cd /content/credit-default-prediction

/content/credit-default-prediction


In [9]:
!python src/download_data.py

Dataset already exists.


# Credit Default Prediction: Preprocessing and Baseline Model

## Objective

This notebook prepares the dataset for machine learning and establishes an initial baseline classification model.

The workflow includes:

- deterministic data cleaning based on findings from EDA,
- separation of features and target,
- stratified train-test splitting,
- preprocessing of numerical and categorical features,
- training of a baseline logistic regression model,
- evaluation using classification metrics appropriate for an imbalanced target.

In [10]:
import pandas as pd
from sklearn.model_selection import train_test_split

In [11]:
df = pd.read_excel(
    "data/raw/default of credit card clients.xls",
    header = 1
)


In [12]:
df_processed = df.copy()

In [13]:
df_processed["EDUCATION"] = df_processed["EDUCATION"].replace([0,5,6],4)
df_processed["EDUCATION"].value_counts()

,count
EDUCATION,
2,14030
1,10585
3,4917
4,468


In [14]:
df_processed['MARRIAGE'] = df_processed["MARRIAGE"].replace(0,3);
df_processed["MARRIAGE"].value_counts()

,count
MARRIAGE,
2,15964
1,13659
3,377


In [15]:
X = df_processed.drop(columns=["ID", "default payment next month"])
y = df_processed["default payment next month"]

In [16]:
X.shape

(30000, 23)

In [17]:
y.shape

(30000,)

In [18]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    stratify=y,
    random_state=42
)

In [19]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(24000, 23)
(6000, 23)
(24000,)
(6000,)


In [20]:
y_train.value_counts(normalize=True) * 100

,proportion
default payment next month,
0,77.879167
1,22.120833


In [23]:
categorical_features = [
    "SEX", "EDUCATION", "MARRIAGE",
    "PAY_0", "PAY_2", "PAY_3",
    "PAY_4", "PAY_5", "PAY_6"
]

numerical_features = [
    "LIMIT_BAL", "AGE",
    "BILL_AMT1", "BILL_AMT2", "BILL_AMT3", "BILL_AMT4", "BILL_AMT5", "BILL_AMT6",
    "PAY_AMT1", "PAY_AMT2", "PAY_AMT3", "PAY_AMT4", "PAY_AMT5", "PAY_AMT6"
]

In [24]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler

In [25]:
preprocessor = ColumnTransformer(
    transformers = [
        ("cat",OneHotEncoder(handle_unknown="ignore"),categorical_features),
        ("num",StandardScaler(),numerical_features)
    ]
)

In [26]:
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression

In [28]:
baseline_model = Pipeline(
    steps =[
        ("preprocessor",preprocessor),
        ("classifier",LogisticRegression(max_iter=1000))
    ]
)

In [29]:
baseline_model.fit(X_train,y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('cat',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['SEX', 'EDUCATION',
                                                   'MARRIAGE', 'PAY_0', 'PAY_2',
                                                   'PAY_3', 'PAY_4', 'PAY_5',
                                                   'PAY_6']),
                                                 ('num', StandardScaler(),
                                                  ['LIMIT_BAL', 'AGE',
                                                   'BILL_AMT1', 'BILL_AMT2',
                                                   'BILL_AMT3', 'BILL_AMT4',
                                                   'BILL_AMT5', 'BILL_AMT6',
                                                   'PAY_AMT1', 'PAY_AMT2',
                                                   'PAY_AMT3', 'PAY_AMT4',
                                                   'PAY_AMT5', 'PAY_AMT6'])])),
                ('classifier', LogisticRegression(max_iter=1000))])

In [30]:
y_pred = baseline_model.predict(X_test)

In [31]:
from sklearn.metrics import (
    confusion_matrix,
    classification_report,
    accuracy_score
)

In [32]:
confusion_matrix(y_test,y_pred)

array([[4435,  238],
       [ 861,  466]])

In [37]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.84      0.95      0.89      4673
           1       0.66      0.35      0.46      1327

    accuracy                           0.82      6000
   macro avg       0.75      0.65      0.67      6000
weighted avg       0.80      0.82      0.79      6000



In [39]:
y_prob = baseline_model.predict_proba(X_test)[:, 1]
y_prob

array([0.15709637, 0.19376916, 0.14088312, ..., 0.06386504, 0.18708702,
       0.11335693])

In [40]:
from sklearn.metrics import roc_auc_score

roc_auc = roc_auc_score(y_test, y_prob)
print(roc_auc)

0.7579977716752476


In [41]:
from sklearn.metrics import average_precision_score

pr_auc = average_precision_score(y_test, y_prob)
print(pr_auc)

0.5275193543057561


### Baseline Model Evaluation

The baseline Logistic Regression model achieved an accuracy of approximately 81.68%. However, recall for the default class was only 35.12%, indicating that a large proportion of actual defaulters were not identified.

The model achieved an ROC-AUC of approximately 0.758, suggesting reasonable ability to distinguish between defaulting and non-defaulting customers across different classification thresholds.

The average precision score was approximately 0.528, which is substantially higher than the positive-class prevalence of approximately 0.221. This indicates that the model provides useful ranking performance for the minority default class.

Overall, the baseline model demonstrates meaningful predictive ability, but its low recall for defaulters indicates that further improvements are required.

In [42]:
balanced_model = Pipeline(
    steps =[
        ("preprocessor",preprocessor),
        ("classifier",LogisticRegression(max_iter=1000,class_weight="balanced"))
    ]
)

In [43]:
balanced_model.fit(X_train,y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('cat',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['SEX', 'EDUCATION',
                                                   'MARRIAGE', 'PAY_0', 'PAY_2',
                                                   'PAY_3', 'PAY_4', 'PAY_5',
                                                   'PAY_6']),
                                                 ('num', StandardScaler(),
                                                  ['LIMIT_BAL', 'AGE',
                                                   'BILL_AMT1', 'BILL_AMT2',
                                                   'BILL_AMT3', 'BILL_AMT4',
                                                   'BILL_AMT5', 'BILL_AMT6',
                                                   'PAY_AMT1', 'PAY_AMT2',
                                                   'PAY_AMT3', 'PAY_AMT4',
                                                   'PAY_AMT5', 'PAY_AMT6'])])),
                ('classifier',
                 LogisticRegression(class_weight='balanced', max_iter=1000))])

In [44]:
y_pred = balanced_model.predict(X_test)

In [45]:
confusion_matrix(y_test,y_pred)

array([[3914,  759],
       [ 590,  737]])

In [47]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.87      0.84      0.85      4673
           1       0.49      0.56      0.52      1327

    accuracy                           0.78      6000
   macro avg       0.68      0.70      0.69      6000
weighted avg       0.79      0.78      0.78      6000



In [51]:
y_prob_balanced = balanced_model.predict_proba(X_test)[:, 1]

In [52]:
roc_auc_score(y_test, y_prob_balanced)

np.float64(0.7585281639252317)

In [53]:
average_precision_score(y_test, y_prob_balanced)

np.float64(0.5250349515868814)

### Effect of Class Weighting

To investigate the effect of class imbalance, a second Logistic Regression model was trained using `class_weight="balanced"`.

Compared with the baseline model, class weighting substantially increased recall for the default class from approximately **35.12% to 55.54%** and improved the F1-score from **45.89% to 52.21%**. However, precision decreased from approximately **66.19% to 49.26%**, while accuracy decreased from **81.68% to 77.52%**.

The ROC-AUC and average precision scores remained similar to those of the baseline model. This suggests that class weighting primarily changed the model's classification behavior rather than substantially improving its overall ability to rank customers by default risk.

In [54]:
y_pred_03 = (y_prob >= 0.30).astype(int)

In [55]:
confusion_matrix(y_test,y_pred_03)

array([[4168,  505],
       [ 692,  635]])

### Effect of Changing the Decision Threshold

The baseline Logistic Regression model initially used the default decision threshold of `0.50`. To investigate the precision-recall trade-off, the threshold was experimentally reduced to `0.30`.

Lowering the threshold increased recall from approximately **35.12% to 47.85%** and improved the F1-score from **45.89% to 51.48%**. However, precision decreased from **66.19% to 55.70%**, and the number of false positives increased.

This demonstrates that the classification threshold can be adjusted according to the relative importance of false negatives and false positives. The value `0.30` is used only as an exploratory threshold and is not considered an optimized threshold.

## Conclusion

This notebook developed the preprocessing workflow and established Logistic Regression as the initial baseline model for credit default prediction.

The baseline model achieved approximately **81.68% accuracy**, but its recall for the default class was only **35.12%**, showing that accuracy alone was not sufficient for evaluating performance on the imbalanced dataset.

Using balanced class weights increased default-class recall to approximately **55.54%** and improved the F1-score, although this resulted in lower precision and accuracy. An exploratory reduction of the classification threshold from `0.50` to `0.30` also demonstrated the trade-off between identifying more defaulters and generating additional false positives.

The ROC-AUC and average precision results indicate that Logistic Regression provides meaningful discriminatory ability, but further model comparison is required.

The next stage will compare multiple classification algorithms using a validation strategy while keeping the final test set separate from model-selection decisions.